### DISCUSSION TITLE
Proposal to change the code style

### DISCUSSION THREAD

**Discussion by nbolton** [deskflow/deskflow#7626] [Development]
Hi all, @sithlord48 proposes we change the code style from LLVM to the style he uses in his projects (proposed in [PR #7624](https://github.com/deskflow/deskflow/pull/7624)). 

> My main things are line length and where opening brackets are.

We should all discuss this before doing so as it's a high impact change.

Our current established [Code Style](https://github.com/deskflow/deskflow/wiki/Code-Style) on the wiki is LLVM:
> CLang (C/C++/Objective-C): .clang-format - LLVM with a few minor tweaks.
> [Clang‐Format Tips & Tricks](https://github.com/deskflow/deskflow/wiki/Clang%E2%80%90Format-Tips-&-Tricks)

**Proposed:**
[`.clang-format`](https://github.com/deskflow/deskflow/blob/e93895addc27504243cdbd391852fe001d7087ce/.clang-format) (style he uses in his personal projects, based on WebKit style)

:warning: **Word of Caution:** Changing the code style to accommodate a contributor's personal preference could have broader implications:

- **Precedent for contributor bias:** Adjusting style for one contributor may set a precedent, increasing friction when others request changes to accommodate their preferences.

- **Impact on current contributors:** Those already accustomed to the LLVM style might find the switch frustrating, affecting productivity and possibly reducing engagement.

Edit: I previously said that the style was the KDE default, which is false; the proposed code style is not widely adopted.

**Comment by nbolton** [comment]
My take: I'm quite happy with LLVM style, but I'm open to change. If we're going to change the code style away from standard LLVM, it should be one that is the most popular to accommodate new contributors.

Example of `.clang-format` files for some of the most popular C++ projects:
- [v8](https://github.com/v8/v8/blob/main/.clang-format) - Google style (with tweaks)
- [googletest](https://github.com/google/googletest/blob/main/.clang-format) - Google style (pure)
- [tensorflow](https://github.com/tensorflow/tensorflow/blob/master/.clang-format) - Google style (pure)
- [godot](https://github.com/godotengine/godot/blob/master/.clang-format) - LLVM style (with tweaks)
- [swift](https://github.com/swiftlang/swift/blob/main/.clang-format) - LLVM style (pure)
- [mysql-server](https://github.com/mysql/mysql-server/blob/trunk/.clang-format) - Google style (verbatim)

**Reply by sithlord48** [reply]
It don't matter what you start with if you set all the options the way you want ...

**Reply by nbolton** [reply]
Perhaps we could stick with the current style, LLVM, which we have used for a while. Existing contributors are used to the current style and LLVM seems to be the 2nd most popular style after Google. WebKit style seems to be one of the least popular, and there could be a good reason for that.

**Reply by gabeklavans** [reply]
Yeah, the more I think about the topic, the more I think it's actually not worth changing the style so late into the project's life. Especially considering the LLVM style is standard enough

**Comment by nbolton** [comment]
I tried this `.clang-format` entry on for size today (from the [KDE `.clang-format` style](https://invent.kde.org/frameworks/extra-cmake-modules/-/blob/master/kde-modules/clang-format.cmake)):
```
# Double the standard 80 column limit to 160, so that Qt `connect` function calls are less likely to wrap.
ColumnLimit: 160
```

After reviewing the format result (406 files are re-formatted), personally, I find the longer lines a bit harder to grok but perhaps I'm just not used to it. Every project I've worked on in the last 25 years has had a character limit of 80 or 100 chars.

Curiously, the [KDevelop project `.clang-format`](https://github.com/KDE/kdevelop/blob/master/.clang-format)  uses 120 columns (overriding KDE's standard 160 limit).

What does everyone else think?

Edit: One more thing to consider, Sonar Cloud (our static code analysis) will treat all changed files as new code and will assert that we have unit tests; it'll force us to lower our code coverage bar and make it hard for us to identify gaps in code coverage for new code for a while.

**Reply by sithlord48** [reply]
There is not a "KDE" Standard for this. every kde project changes it for their own needs

I think i can fit ~200 characters per line on even small screens. Most computers now are using a 16:9 / 16:10 format they are WAY wider then monitors where when that 80 column wrapping was common. I trust ppl will make sane code breaks and the limit will only end up breakin lines the author intended to be a single line.

**Reply by DominikDeak** [reply]
> Curiously, the [KDevelop project `.clang-format`](https://github.com/KDE/kdevelop/blob/master/.clang-format) uses 120 columns (overriding KDE's standard 160 limit).

There is nothing wrong with using extended columns. These days it's rare to see someone developing within a 80 column terminal screen, and most displays are huge. So wider the better.

As for the rest of the formatting stuff, just pick one of the industry standard ones and stick with it. Doesn't matter that much, so long it's consistent. And with clang format that consistency is easy to automate and enforce.

**Reply by nbolton** [reply]
My main concern is not screen size. Certainly, the 80-column limit is a legacy standard, but it’s still relevant for a few practical reasons:

1.	Readability: Shorter lines are easier to read and comprehend. When code stretches too far horizontally, it can become difficult to follow.

2.	Code Comparison and Review: Many side-by-side comparison tools (diff views) struggle with wider lines, making it hard to review changes. Sticking to 80 columns ensures code fits neatly in these views without excessive horizontal scrolling.

3.	Accessibility: Developers using smaller screens or split-view editors (e.g., two files open side by side) benefit from shorter lines. It minimizes scrolling and helps retain focus. Also consider that not everyone uses a horizontal screen.

4.	Consistent Formatting: Limiting columns encourages more disciplined formatting, with shorter function names, smaller blocks, and fewer levels of indentation, which generally leads to cleaner, more maintainable code.

A common argument against this could be "But you should let the developer decide how much of the line to use." The problem with this argument is that Clang format will fill the line to the max column width if it can, so even if you don't want it to stretch the whole width, the formatter will force this. This is likely why the default is 80 for all but one of the Clang format styles (with WebKit being the outlier at 120).

**Comment by nbolton** [comment]
> where opening brackets are

@sithlord48 Do you mean that you prefer the [WebKit](https://www.webkit.org/code-style-guidelines/) style where curly braces are on the next line?
```
void foo()
{
}
```

It seems that the most popular styles, [LLVM](https://llvm.org/docs/CodingStandards.html) and [Google](https://google.github.io/styleguide/cppguide.html) have curlies on the same line.
```
void foo() {
}
```

My personal preference is strongly with LLVM and Google.

**Reply by sithlord48** [reply]
It depends on what foo is for classes
```
void foo ()
{

}
```

for if 

```
if {

} else {

}
```

The current clang format has all the line options explicitly set.

**Comment by gabeklavans** [comment]
My take is if there are certain devs that are going to develop way more on this project, then they can set the style. @sithlord48 it seems you have a lot of momentum. It also seems you and @nbolton are the main project leaders right now. If you're in agreement, and the style guide is somewhat modern, I say make it so. My only input would be to try to use a well-defined standard so that we don't spend time messing with it. I'm not sure how well defined/widely used the KDE formatting rules are. I'm more inclined towards something like https://google.github.io/styleguide/cppguide.html as it's more generally used and is integrated into formatting/linting tools. Also, Deskflow isn't really a KDE project, but once again if the bulk of the work will be done by a KDE/QT dev, then go for it.

**Reply by nbolton** [reply]
The bulk of the work will likely be done by non-KDE devs.

My personal preference is the current style, LLVM. I prefer this over the KDE/WebKit style, but it's subjective so I'm happy to go with the majority. I have worked with many code styles over the years so I'm pretty flexible. However, I do think going for a 160 column limit would have adverse effects (as I pointed out in another thread). I'm open to 100 chars and I have worked on projects where this just about works, but there were some minor gripes with side-by-side diffs.

We could do with a few more regular contributors to weigh in here to help us decide if we should take the lowest friction route (stick with LLVM style) or make the high impact change to switch code styles over to KDE/WebKit.

**Comment by nbolton** [comment]
Surfacing https://github.com/deskflow/deskflow/pull/7624#issuecomment-2413614009:
> I still want readable code the linter settings as have way to short of line and keep forcing lines be full length etc.. Honestly just get rid of it

I understand your frustration with Clang-Format. It's not perfect. But getting rid of it entirely? That’s just trading one set of problems for another. The consistency it brings is hard to beat. 

When code formatting becomes a manual task, it’s easy to get bogged down by personal preferences and inconsistencies. Clang-Format takes the cognitive load off by automating that process, and while it might not always format the code as you'd do it yourself, the trade-off is fewer distractions and more focus on the actual code logic. Fighting it can become frustrating and waste valuable time. Adopting the tool and adapting to its quirks is often the best path for long-term productivity.

Edit: I wrote an article to help with Clang-Format frustrations: [Clang‐Format Tips & Tricks](https://github.com/deskflow/deskflow/wiki/Clang%E2%80%90Format-Tips-&-Tricks)

**Reply by nbolton** [reply]
@DominikDeak @sithlord48 @gabeklavans Any thoughts on this? I want to get the discussion wrapped up so we can move past the code style issue as I think it could be sapping our energy a bit.

On this topic, I have started a wiki article to help people struggling with Clang-Format (inspired by KDE's docs):
- [Clang‐Format Tips & Tricks](https://github.com/deskflow/deskflow/wiki/Clang%E2%80%90Format-Tips-&-Tricks)

**Reply by gabeklavans** [reply]
I agree. Forced format is better than no format. Maybe some rules feel legacy nowadays, like the column width, but Deskflow is not a *new* project, and like Nick has said, lots of people old and new will be working on this, so there’s no particular reason to change the whole codebase up for formatting reasons.

My personal anecdote is I haven’t had any noticeable issue reading the code so far. Feels the same as reading C++ in any other codebase :)

**Comment by OwenMPhillips** [comment]
I share some of the same concerns that have been raised already such as having to lower the code coverage bar.

There is no perfect code style. In my opinion, unless there is a significant advantage to changing it so late into the project, it is not worth the effort. 

Side note: I'm on an ultrawide monitor and only ever stretch my IDE to use half the screen. I prefer 100 characters per line.

Disclaimer: I work for Symless.

**Reply by sithlord48** [reply]
I am not saying we need to use 300 characters for all lines

what i am trying to say is the forcing of 100 or 80 or 160 len lines is not good.  The linter just pads them that is for sure part of the problem with linters

in c++ we can get long namespaces etc i trust devs to do the right thing w line length

**Comment by nbolton** [comment]
I was hoping for more people to be involved in the discussion, and I hope that I'm not wrapping this up too soon.

It looks like the consensus--between all that have contributed to this discussion--is that we should keep the current code style (LLVM) and Clang linter as documented on the wiki: [Code Style](https://github.com/deskflow/deskflow/wiki/Code-Style)

**Reply by nbolton** [reply]
Looks like we don't have a unanimous vote, which would have been my preference. So, let's try to find a middle ground.

**Comment by sithlord48** [comment]
I've updated the style proposed to be based on LLVM . 120 column limits Please go check it out #7624

**Reply by nbolton** [reply]
While looking for a reason to use 120 inline and be consistent, I found that before applying `clang-format` to the repo, there are examples of files that are up to 120 columns long. I mistakingly thought that the code style was limited columns 80 but that's not the case. I dug through the git history and gathered some statistics for you all.

For `.cpp` and `.h` files:
```
Files with max column width <80: 346
Files with max column width 80-100: 158
Files with max column width 100-120: 85
```

The majority are definitely <80 columns, but I think there are enough examples of files that are up to 120 to justify that as a precedent.

For example, this line is 120 chars exactly:
```
ClientProxyUnknown::ClientProxyUnknown(synergy::IStream* stream, double timeout, Server* server, IEventQueue* events) :
```

Guess who wrote that line! :slightly_smiling_face: 

![image](https://github.com/user-attachments/assets/278c39cc-7870-4754-85f1-f6c68af7c61b)

So, I can't really say that 120 is unreasonable as it has been a precedent for at least a decade now.

And... braces are on the next line.
```
void
Log::print(const char* file, int line, const char* fmt, ...)
{
```

Looks like your new proposed style is more inline with the OG code style than the current style.

**Comment by nbolton** [comment]
PR merged! #7624
